libbeat/common/reload/reload.go

```diff
@@ -0,0 +1,105 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package reload
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+var register = newRegistry()
+
+// ConfigWithMeta holds a pair of common.Config and optional metadata for it
+type ConfigWithMeta struct {
+	// Config to store
+	Config *common.Config
+
+	// Meta data related to this config
+	Meta *common.MapStrPointer
+}
+
+// ReloadableList provides a method to reload the configuration of a list of entities
+type ReloadableList interface {
+	Reload(configs []*ConfigWithMeta) error
+}
+
+// Reloadable provides a method to reload the configuration of an entity
+type Reloadable interface {
+	Reload(config *ConfigWithMeta) error
+}
+
+type registry struct {
+	sync.RWMutex
+	confsLists map[string]ReloadableList
+	confs      map[string]Reloadable
+}
+
+func newRegistry() registry {
+	return registry{
+		confsLists: make(map[string]ReloadableList),
+		confs:      make(map[string]Reloadable),
+	}
+}
+
+// MustRegister declares a reloadable object
+func MustRegister(name string, r Reloadable) {
+	register.Lock()
+	defer register.Unlock()
+
+	if r == nil {
+		panic("Got a nil object")
```

Comment:
What about returning an error instead?

Reply:
All the register calls should happen while initializing the beat, if an error happens here it means we did something really wrong, it would be interesting to know that early, I think we follow this pattern in many other registries?

Reply:
ok, lets follow the common pattern.

Reply:
It's good practice to offer both workflow.

Reply:
I added another commit to offer both in a non-global fashion